feat: add more validation for custom policy file

Abhra303 opened the pull request:
Fixes #519 

These validations are - 

> Identifier:
      a. verify each rule has a unique identifier
      b. verify that each identifier has a corresponding rule to run against
verify that exactly 1 policy has it's isDefault value set to true
jsonSchema field - verify that the schema is valid
schema field - verify that the schema is valid

royhadad commented on bl/policy/policy_factory.go:
```diff
@@ -37,13 +38,20 @@ func CreatePolicy(policies *cliClient.EvaluationPrerunPolicies, policyName strin
+
+				err = fmt.Errorf("invalid policy file: multiple policies are set to default")
+				return Policy{}, err
+			} else if policyName == "" && policy.IsDefault {
```
if the user provides a policyName (`--policy` flag), this check will never detect the duplication because `policyName != ""`

royhadad commented on bl/policy/policy_factory.go:
```diff
@@ -37,13 +38,20 @@ func CreatePolicy(policies *cliClient.EvaluationPrerunPolicies, policyName strin
-				break
-			} else if policy.Name == policyName {
+				sawDefault = true
+			} else if policy.Name == policyName && !sawDefault {
```
This addition to the `if` statement also doesn't seem logical to me

royhadad commented on bl/policy/policy_factory.go:
```diff
@@ -7,6 +7,7 @@ import (
 
 	"github.com/datreeio/datree/pkg/cliClient"
 	internal_policy "github.com/datreeio/datree/pkg/policy"
+	"github.com/xeipuuv/gojsonschema"
```
We're gonna need to move all of the validations into `/pkg/policy/policy.go` - `validatePoliciesYaml()`

Abhra303 replied:
Actually, at first I was thinking of adding these validations to `validatePoliciesYaml()` function. But I thought, adding the validation logic to that function would need lots of `yaml.unmarshal` and `for` loops (to check schemas, identifier uniqueness etc.) which we already have in `policy_factory.go` (e.g. [`populateRules()`](https://github.com/datreeio/datree/blob/359680591f6cce4ebc97f0f8e8462b1b2d431c17/bl/policy/policy_factory.go#L69)). So whenever user try to use `--policy-config` , similar operations would run two times (one for validation check and another is for parsing the policies).

We cannot bypass the `populateRules()` function because [`policies`](https://github.com/datreeio/datree/blob/359680591f6cce4ebc97f0f8e8462b1b2d431c17/cmd/test/main.go#L286) given by `evaluationPrerunDataResp.PoliciesJson` need this function. so I thought it would be simpler and would avoid running similar operations two times.

What's your view on it?

royhadad replied:
In my view, in 95% of cases - maintainability > speed (unless there is an actual performance issue)
This may be a bit faster but much harder to maintain, the separation of steps makes it easier to wrap your head around it and make changes to the code later 😊 

step 1 - validate the input
step 2 - use the input (once we know it's clean)

Abhra303 replied:
Thanks for clearing my doubt! Will make the necessary changes.

royhadad commented on bl/policy/policy_factory.go:
```diff
@@ -7,6 +7,7 @@ import (
 
 	"github.com/datreeio/datree/pkg/cliClient"
 	internal_policy "github.com/datreeio/datree/pkg/policy"
+	"github.com/xeipuuv/gojsonschema"
```
We'll need matching test cases for the new validations under `/pkg/policy/policy_test.go`

royhadad commented on bl/policy/policy_factory.go:
```diff
@@ -102,6 +123,27 @@ func populateRules(policyRules []cliClient.Rule, customRules []*cliClient.Custom
+	schemaLoader := gojsonschema.NewStringLoader(string(jsonContent))
+	_, err = gojsonschema.NewSchemaLoader().Compile(schemaLoader)
+	if err != nil {
+		return fmt.Errorf("invalid policy file: rule schema error: %s", err.Error())
```
We would like to add the rule identifier in the error message, this way the user will know in which custom rule the error occurred.

royhadad commented on bl/policy/policy_factory.go:
```diff
@@ -112,14 +154,22 @@ func getDefaultRuleByIdentifier(defaultRules []*internal_policy.DefaultRuleDefin
+				// If we are here that means we have found another
+				// identifier with the same name. So, throw error
+
+				err := fmt.Errorf("invalid policy file: identifiers are not unique")
```
every function should do one thing, this function has 2 responsibilities:
1. get the customRule
2. validate that every customRule has a unique identifier 

On another note, the validation for rules identifiers uniqueness should be across all default rules **and** custom rules

royhadad commented on bl/policy/policy_factory.go:
```diff
@@ -102,6 +123,27 @@ func populateRules(policyRules []cliClient.Rule, customRules []*cliClient.Custom
 }
 
+func checkRuleSchemaExist(customRule *cliClient.CustomRule, identifier string) error {
+	if customRule.Schema == nil && customRule.JsonSchema == "" {
```
this should fail in the schema validation step, no need for a special check beforehand

royhadad reviewed (changes requested):
@Abhra303 We greatly appreciate the contribution! I have left some crucial comments that need to be addressed before we can merge it 😊 

I suggest you start with this one:
https://github.com/datreeio/datree/pull/605#discussion_r867909100

royhadad commented:
Hi @Abhra303!
just wanted to let you know that I just added the infra for testing the `validatePoliciesYaml` function!
all you have to do is add a test case in `/pkg/policy/validatePoliciesYaml_test.go` and add the case file to the `validatePoliciesYamlFixtures` folder

https://github.com/datreeio/datree/blob/main/pkg/policy/validatePoliciesYaml_test.go

Abhra303 commented:
Thanks @royhadad 😊

Will try to submit the changes by tomorrow :)

Abhra303 committed:
feat: add more validation checks for policy file

When user uses `--policy-config`, datree first validates the policy
file using a jsonSchema file. But it does not check things like
identifier uniqueness, single default policy check etc.

Add those check in `validatePoliciesYaml` package and also move
`defaultRules` specific struct and functions to a dedicated package
called `defaultRules` for better maintainability.

Abhra303 committed:
Merge branch 'main' into ISSUE#519_add_validation

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml_test.go:
```diff
@@ -53,11 +71,17 @@ func TestValidatePoliciesYaml(t *testing.T) {
 	assertValidationResult(t, missingPolicyRules, "./test_fixtures/missingPolicyRules.yaml", nil)
 	assertValidationResult(t, missingApiVersion, "./test_fixtures/missingApiVersion.yaml", errors.New("found errors in policies file ./test_fixtures/missingApiVersion.yaml:\n(root): missing properties: 'apiVersion'"))
 	assertValidationResult(t, missingPolicyName, "./test_fixtures/missingPolicyName.yaml", errors.New("found errors in policies file ./test_fixtures/missingPolicyName.yaml:\n(root)/policies/0: missing properties: 'name'"))
+	assertValidationResult(t, missingSchema, "./test_fixtures/missingSchema.yaml", errors.New("invalid policy file: rule identifier \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has no associated schema"))
```
We want the error to be:
"found errors in policies file ./test_fixtures/missingSchema.yaml:\n(root)/customRules/1: Exactly one of [schema,jsonSchema] should be defined per custom rule"

note that the "1" in "customRules/1" is dynamic

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml_test.go:
```diff
@@ -53,11 +71,17 @@ func TestValidatePoliciesYaml(t *testing.T) {
 	assertValidationResult(t, missingApiVersion, "./test_fixtures/missingApiVersion.yaml", errors.New("found errors in policies file ./test_fixtures/missingApiVersion.yaml:\n(root): missing properties: 'apiVersion'"))
 	assertValidationResult(t, missingPolicyName, "./test_fixtures/missingPolicyName.yaml", errors.New("found errors in policies file ./test_fixtures/missingPolicyName.yaml:\n(root)/policies/0: missing properties: 'name'"))
+	assertValidationResult(t, missingSchema, "./test_fixtures/missingSchema.yaml", errors.New("invalid policy file: rule identifier \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has no associated schema"))
+	assertValidationResult(t, multipleDefaultPolicy, "./test_fixtures/multipleDefaultPolicy.yaml", errors.New("invalid policy file: multiple policies are set to default"))
```
We want the error to be:
"found errors in policies file ./test_fixtures/multipleDefaultPolicy.yaml:\n(root)/policies: Should have exactly one policy set as default"

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml_test.go:
```diff
@@ -53,11 +71,17 @@ func TestValidatePoliciesYaml(t *testing.T) {
 	assertValidationResult(t, customRuleMissingName, "./test_fixtures/customRuleMissingName.yaml", errors.New("found errors in policies file ./test_fixtures/customRuleMissingName.yaml:\n(root)/customRules/0: missing properties: 'name'"))
 	assertValidationResult(t, customRuleMissingDefaultMessageOnFailure, "./test_fixtures/customRuleMissingDefaultMessageOnFailure.yaml", errors.New("found errors in policies file ./test_fixtures/customRuleMissingDefaultMessageOnFailure.yaml:\n(root)/customRules/0: missing properties: 'defaultMessageOnFailure'"))
 	assertValidationResult(t, customRuleValidSchema, "./test_fixtures/customRuleValidSchema.yaml", nil)
+	assertValidationResult(t, customRuleInvalidSchema, "./test_fixtures/customRuleInvalidSchema.yaml", errors.New("invalid policy file: rule \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has schema error: has a primitive type that is NOT VALID -- given: /arrayi/ Expected valid values are:[array boolean integer number null object string]"))
```
We want the error to be:
"found errors in policies file ./test_fixtures/customRuleInvalidSchema.yaml:\n(root)/customRules/1/schema: has a primitive type that is NOT VALID -- given: /arrayi/ Expected valid values are:[array boolean integer number null object string]"

note that the "1" in "customRules/1" is dynamic

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml_test.go:
```diff
@@ -53,11 +71,17 @@ func TestValidatePoliciesYaml(t *testing.T) {
 	assertValidationResult(t, customRuleMissingDefaultMessageOnFailure, "./test_fixtures/customRuleMissingDefaultMessageOnFailure.yaml", errors.New("found errors in policies file ./test_fixtures/customRuleMissingDefaultMessageOnFailure.yaml:\n(root)/customRules/0: missing properties: 'defaultMessageOnFailure'"))
 	assertValidationResult(t, customRuleValidSchema, "./test_fixtures/customRuleValidSchema.yaml", nil)
+	assertValidationResult(t, customRuleInvalidSchema, "./test_fixtures/customRuleInvalidSchema.yaml", errors.New("invalid policy file: rule \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has schema error: has a primitive type that is NOT VALID -- given: /arrayi/ Expected valid values are:[array boolean integer number null object string]"))
+	assertValidationResult(t, customRuleInvalidJsonSchema, "./test_fixtures/customRuleInvalidJsonSchema.yaml", errors.New("invalid policy file: rule \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has schema error: Invalid type. Expected: array of schemas, given: definitions"))
```
We want the error to be:
"found errors in policies file ./test_fixtures/customRuleInvalidJsonSchema.yaml:\n(root)/customRules/1/schema: Invalid type. Expected: array of schemas, given: definitions"

note that the "1" in "customRules/1" is dynamic

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml_test.go:
```diff
@@ -53,11 +71,17 @@ func TestValidatePoliciesYaml(t *testing.T) {
 	assertValidationResult(t, customRuleValidSchema, "./test_fixtures/customRuleValidSchema.yaml", nil)
+	assertValidationResult(t, customRuleInvalidSchema, "./test_fixtures/customRuleInvalidSchema.yaml", errors.New("invalid policy file: rule \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has schema error: has a primitive type that is NOT VALID -- given: /arrayi/ Expected valid values are:[array boolean integer number null object string]"))
+	assertValidationResult(t, customRuleInvalidJsonSchema, "./test_fixtures/customRuleInvalidJsonSchema.yaml", errors.New("invalid policy file: rule \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has schema error: Invalid type. Expected: array of schemas, given: definitions"))
+	assertValidationResult(t, customRuleIdentifierNotUnique1, "./test_fixtures/customRuleIdentifierNotUnique1.yaml", errors.New("invalid policy file: identifier \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" is used in atleast two rules"))
```
We want the error to be:
"found errors in policies file ./test_fixtures/customRuleIdentifierNotUnique1.yaml:\n(root)/customRules: identifier \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" is used in more than one custom rule"

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml_test.go:
```diff
@@ -53,11 +71,17 @@ func TestValidatePoliciesYaml(t *testing.T) {
+	assertValidationResult(t, customRuleInvalidSchema, "./test_fixtures/customRuleInvalidSchema.yaml", errors.New("invalid policy file: rule \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has schema error: has a primitive type that is NOT VALID -- given: /arrayi/ Expected valid values are:[array boolean integer number null object string]"))
+	assertValidationResult(t, customRuleInvalidJsonSchema, "./test_fixtures/customRuleInvalidJsonSchema.yaml", errors.New("invalid policy file: rule \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has schema error: Invalid type. Expected: array of schemas, given: definitions"))
+	assertValidationResult(t, customRuleIdentifierNotUnique1, "./test_fixtures/customRuleIdentifierNotUnique1.yaml", errors.New("invalid policy file: identifier \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" is used in atleast two rules"))
+	assertValidationResult(t, customRuleIdentifierNotUnique2, "./test_fixtures/customRuleIdentifierNotUnique2.yaml", errors.New("invalid policy file: identifier \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" used atleast two times in policy \"Default\""))
```
We want the error to be:
"found errors in policies file ./test_fixtures/customRuleIdentifierNotUnique2.yaml:\n(root)/policies/0/rules: identifier "PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION" is used more than once in policy"

note that the "0" in "policies/0" is dynamic

royhadad commented on pkg/validatePoliciesYaml/test_fixtures/customRuleIdentifierNotUnique2.yaml:
```diff
@@ -0,0 +1,17 @@
+apiVersion: v1
```
maybe rename this file to:
`duplicateRuleIdentifier.yaml`

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml.go:
```diff
@@ -30,5 +34,128 @@ func ValidatePoliciesYaml(content []byte, policyYamlPath string) error {
+			sawDefault = true
+		}
+	}
+	return nil
```
what if the user provides 0 policies set to default? we also want it to fail, I suggest adding a test case for that :)

Abhra303 replied:
Good catch! Will add it 👍

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml.go:
```diff
@@ -30,5 +34,128 @@ func ValidatePoliciesYaml(content []byte, policyYamlPath string) error {
+		identifier := item.UniqueName
+
+		if propertyValuesExistenceMap[identifier] {
+			return fmt.Errorf("invalid policy file: a default rule with same identifier \"%s\" already exists", identifier)
```
please match this error message to match the updated structure of the other error messages and consider adding a test case :)

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml.go:
```diff
@@ -30,5 +34,128 @@ func ValidatePoliciesYaml(content []byte, policyYamlPath string) error {
+			schema := rule.Schema
+			content, err = json.Marshal(schema)
+			if err != nil {
+				return err
```
please match this error message structure to the other updated error messages and consider adding a test case :)

Abhra303 replied:
Hello @royhadad , I improved the error message but I am struggling to find the test case for it. When will `json.Marshal` throw error for a portion of a valid yaml content?

royhadad replied:
you are correct, this should never happen

royhadad commented on pkg/validatePoliciesYaml/validatePoliciesYaml_test.go:
```diff
@@ -53,11 +71,17 @@ func TestValidatePoliciesYaml(t *testing.T) {
 	assertValidationResult(t, missingPolicyRules, "./test_fixtures/missingPolicyRules.yaml", nil)
 	assertValidationResult(t, missingApiVersion, "./test_fixtures/missingApiVersion.yaml", errors.New("found errors in policies file ./test_fixtures/missingApiVersion.yaml:\n(root): missing properties: 'apiVersion'"))
 	assertValidationResult(t, missingPolicyName, "./test_fixtures/missingPolicyName.yaml", errors.New("found errors in policies file ./test_fixtures/missingPolicyName.yaml:\n(root)/policies/0: missing properties: 'name'"))
+	assertValidationResult(t, missingSchema, "./test_fixtures/missingSchema.yaml", errors.New("invalid policy file: rule identifier \"PODDISRUPTIONBUDGET_DENY_ZERO_VOLUNTARY_DISRUPTION\" has no associated schema"))
```
missing test case & check for case where the customRule has both `schema` & `jsonSchema` keys, should get the same error message:

"found errors in policies file ./test_fixtures/missingSchema.yaml:\n(root)/customRules/1: Exactly one of [schema,jsonSchema] should be defined per custom rule"

Abhra303 replied:
Thanks for pointing out :)

royhadad reviewed (changes requested):
Awesome work! The code is so readable and easy to understand 😊 
most of the changes are adjustments to the error messages and some missing checks, this should be the last iteration.
I suggest you work through the adjustments [TDD](https://en.wikipedia.org/wiki/Test-driven_development) style, changing/adding 1 test case at a time and then making it pass

royhadad commented:
Hey, I think this is missing another validation & test case, where the policy rule identifier is neither a custom rule nor default rule (1.b)
![Screen Shot 2022-05-15 at 13 57 05](https://user-images.githubusercontent.com/39004075/168469578-80e67edf-11f5-4f0b-861d-e5a785427216.png)

Abhra303 committed:
Merge branch 'main' into ISSUE#519_add_validation

Abhra303 committed:
fix: add missing checks and test cases

Abhra303 commented:
Hello @royhadad , I hopefully addressed all the review messages in the latest change.Kindly inform me, if you find something missing here :)

royhadad committed:
Merge branch 'main' into ISSUE#519_add_validation

royhadad reviewed (approved):
Awesome work! merging this 😊😊😊
